### PR TITLE
4.1 - Added AggregateExpression for aggregate functions

### PR DESCRIPTION
**Pull request by othercorey**
https://github.com/cakephp/cakephp/issues/14080

This is the first part of window function support. This adds the necessary classes and API in Query to start building.

The AggregateExpression class will support the extended API for window functions.  We can also add full support for extended aggregate expression clauses in the future.

~The existing aggregate functions in FunctionsBuilder will still work, but they will not be extendable. Eventually, we can deprecate them in favor of the extensible versions.~

As there is no new behavior in this copy, the unit tests are just copying the FunctionExpression tests.

### CODE REVIEW

File: src/Database/AggregatesBuilder.php [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+class AggregatesBuilder
+{
+    /**
+     * Returns a AggregateExpression representing a call to SQL SUM function.
+     *
+     * @param mixed $expression the function argument
+     * @param array $types list of types to bind to the arguments
+     * @return \Cake\Database\Expression\AggregateExpression
+     */
+    public function sum($expression, $types = []): AggregateExpression
```
Comment (markstory) [inline src/Database/AggregatesBuilder.php]: Would the planned window related methods be available on the sum() expression?

**Review by markstory — COMMENTED** [review]
Do we need to separate the builders for generic functions and aggregates? It sounded like we would be able to expose the required windowing features on a the window expression class, without requiring more builders.

File: src/Database/AggregatesBuilder.php (hunk 1 above)
Comment (othercorey) [reply src/Database/AggregatesBuilder.php]: Yes, any supported aggregate function can be extend to a window function.

Even without that, we can add FILTER support to this interface for non-window functions.

**Comment by othercorey** [comment]
> Do we need to separate the builders for generic functions and aggregates? It sounded like we would be able to expose the required windowing features on a the window expression class, without requiring more builders.

Sure, that's up for debate.  The separate builder does two things:

1. Ensure an AggregateExpression is returned and typed
2. Avoid user mistakes when mixing non-aggregate functions with aggregate clauses and window clauses.

**Comment by markstory** [comment]
> Ensure an AggregateExpression is returned and typed

We could do this with the current function builder too if each window-able function was typed to the window expression type.

How do you see the API working for this approach on window functions? How would someone use them in a query?

**Comment by othercorey** [comment]
> We could do this with the current function builder too if each window-able function was typed to the window expression type.

How do we support the arbitrary function building with that?  That's the main reason I switched to this.  I'm not going to wrap every aggregate and window function.

> 
> How do you see the API working for this approach on window functions? How would someone use them in a query?

There are 2 common approaches I had in mind.  There are some caveats with the vendor implementations and nuances on customization to the window when in query vs column and with duplicate names/etc.  I'm hoping to hit the common stuff.

```
// set existing window
$sum = $query->aggregate()->sum('field')->window('name' or $window);
```

```
// build window
$sum = $query->aggregate()->sum('field');
$sum
    ->window()
    ->partition()
    ->order()
    ->frame();
```

You could add a separate window to the query:

```
$query->window('name', $window);
```

**Comment by markstory** [comment]
> How do we support the arbitrary function building with that? That's the main reason I switched to this. I'm not going to wrap every aggregate and window function.

Ok. Using the function builder would require using the aggregate expression class in specific functions. We could also have a `window()` (bad name) as a way to support other function names.

With having another builder I would be concerned with the two sum() methods being different.

**Comment by othercorey** [comment]
> With having another builder I would be concerned with the two sum() methods being different.

Maybe there is another approach.  I don't think having `aggregate()` just for the call builder will be clean.  Users won't understand why `func()` for pre-defined is different from `aggregate()`.

To avoid diverging implementations/double implementation, why don't we give `FunctionsBuilder` an `AggregatesBuilder` instance that it uses to build the existing aggregate functions wrapped by `FunctionBuilder`.  That way they both use the same implementation and require the same interface and types.

This would also fix the issue where current `func()->sum()` expressions would not be extendable to window functions.

**Comment by markstory** [comment]
> Maybe there is another approach. I don't think having aggregate() just for the call builder will be clean. Users won't understand why func() for pre-defined is different from aggregate().

Are you referring to the `__call` method? Could we have a `buildAggregate()` that returns an `AggregateExpression` instead of the vanilla `FunctionExpression`? I don't think we need to have `__call` support for aggregates if we have a build method. We can add more concrete methods as requests come in.

**Comment by othercorey** [comment]
> I don't think we need to have `__call` support for aggregates if we have a build method. We can add more concrete methods as requests come in.

That feels like I'm going to break someone.  

What about the option of just having `FunctionExpression` support everything and leave it up to the user if they added the wrong option for a non-aggregate function?

**Comment by markstory** [comment]
> What about the option of just having FunctionExpression support everything and leave it up to the user if they added the wrong option for a non-aggregate function?

Is the scenario you're concerned about

```php
$query->func()->customaggregate()->window()
```

Not working? Wouldn't the `__call` return type let the developer know that this would fail as the return of `__call` is a `FunctionExpression`? We have concrete functions on the `FunctionsBuilder` for aggregate functions (`sum`, `count` etc). Those functions could return an `AggregateFunctionExpression` which supports the new windowing features. If users need more functions with windowing we can either add more concrete methods to `FunctionsBuilder` or those people can use 

```php
$query->func()->buildAggregate('customaggregate', ...)->window()
```

**Comment by othercorey** [comment]
- Removed `AggregatesBuilder`
- Updated `FunctionsBuilder` to return `AggregateExpression` for existing aggregates
- Added `FunctionsBuilder::aggregate()` to support building arbitrary aggregates.

**Review by markstory — APPROVED** [review]
Looks good to me. We can expand this in future pull requests :+1:

**Pull request merged**